### PR TITLE
fix(anta): Fix ResultManager statistics

### LINKED ISSUE #948: fix: `anta nrfu table --group-by` is broken
```
mtache@cactus:/mnt/lab/projects/anta-demo$ anta nrfu -c 1-network-tests/nrfu.yml table --group-by device
╭────────────────────── Settings ──────────────────────╮
│ - ANTA Inventory contains 6 devices (AsyncEOSDevice) │
│ - Tests catalog contains 11 tests                    │
╰──────────────────────────────────────────────────────╯

[12:27:19] INFO     Preparing ANTA NRFU Run ...                                                                                                                                                                                                                                                                                                                                                                                                                            tools.py:293
           INFO     Connecting to devices ...                                                                                                                                                                                                                                                                                                                                                                                                                              tools.py:293
           INFO     Connecting to devices completed in: 0:00:00.030.                                                                                                                                                                                                                                                                                                                                                                                                       tools.py:301
           INFO     Preparing the tests ...                                                                                                                                                                                                                                                                                                                                                                                                                                tools.py:293
           INFO     Preparing the tests completed in: 0:00:00.001.                                                                                                                                                                                                                                                                                                                                                                                                         tools.py:301
           INFO     --- ANTA NRFU Run Information ---                                                                                                                                                                                                                                                                                                                                                                                                                     runner.py:276
                    Number of devices: 6 (6 established)                                                                                                                                                                                                                                                                                                                                                                                                                               
                    Total number of selected tests: 50                                                                                                                                                                                                                                                                                                                                                                                                                                 
                    Maximum number of open file descriptors for the current ANTA process: 16384                                                                                                                                                                                                                                                                                                                                                                                        
                    ---------------------------------                                                                                                                                                                                                                                                                                                                                                                                                                                  
           INFO     Preparing ANTA NRFU Run completed in: 0:00:00.034.                                                                                                                                                                                                                                                                                                                                                                                                     tools.py:301
           INFO     Running ANTA tests ...                                                                                                                                                                                                                                                                                                                                                                                                                                 tools.py:293
           INFO     Running ANTA tests completed in: 0:00:00.125.                                                                                                                                                                                                                                                                                                                                                                                                          tools.py:301
           INFO     Cache statistics for 'spine1': 0 hits / 8 command(s) (0.00%)                                                                                                                                                                                                                                                                                                                                                                                           runner.py:75
           INFO     Cache statistics for 'spine2': 0 hits / 8 command(s) (0.00%)                                                                                                                                                                                                                                                                                                                                                                                           runner.py:75
           INFO     Cache statistics for 'leaf1': 1 hits / 10 command(s) (10.00%)                                                                                                                                                                                                                                                                                                                                                                                          runner.py:75
           INFO     Cache statistics for 'leaf2': 1 hits / 10 command(s) (10.00%)                                                                                                                                                                                                                                                                                                                                                                                          runner.py:75
           INFO     Cache statistics for 'leaf3': 1 hits / 10 command(s) (10.00%)                                                                                                                                                                                                                                                                                                                                                                                          runner.py:75
           INFO     Cache statistics for 'leaf4': 1 hits / 10 command(s) (10.00%)                                                                                                                                                                                                                                                                                                                                                                                          runner.py:75
  • Running NRFU Tests...100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50/50 • 0:00:00 • 0:00:00

                                            Summary per device                                            
┏━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Device ┃ # of success ┃ # of skipped ┃ # of failure ┃ # of errors ┃ List of failed or error test cases ┃
┡━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ leaf1  │ 0            │ 0            │ 0            │ 0           │                                    │
├────────┼──────────────┼──────────────┼──────────────┼─────────────┼────────────────────────────────────┤
│ leaf2  │ 0            │ 0            │ 0            │ 0           │                                    │
├────────┼──────────────┼──────────────┼──────────────┼─────────────┼────────────────────────────────────┤
│ leaf3  │ 0            │ 0            │ 0            │ 0           │                                    │
├────────┼──────────────┼──────────────┼──────────────┼─────────────┼────────────────────────────────────┤
│ leaf4  │ 0            │ 0            │ 0            │ 0           │                                    │
├────────┼──────────────┼──────────────┼──────────────┼─────────────┼────────────────────────────────────┤
│ spine1 │ 0            │ 0            │ 0            │ 0           │                                    │
├────────┼──────────────┼──────────────┼──────────────┼─────────────┼────────────────────────────────────┤
│ spine2 │ 0            │ 0            │ 0            │ 0           │                                    │
└────────┴──────────────┴──────────────┴──────────────┴─────────────┴────────────────────────────────────┘
```


```
mtache@cactus:/mnt/lab/projects/anta-demo$ anta nrfu -c 1-network-tests/nrfu.yml table --group-by test
╭────────────────────── Settings ──────────────────────╮
│ - ANTA Inventory contains 6 devices (AsyncEOSDevice) │
│ - Tests catalog contains 11 tests                    │
╰──────────────────────────────────────────────────────╯

[12:28:02] INFO     Preparing ANTA NRFU Run ...                                                                                                                                                                                                                                                                                                                                                                                                                            tools.py:293
           INFO     Connecting to devices ...                                                                                                                                                                                                                                                                                                                                                                                                                              tools.py:293
           INFO     Connecting to devices completed in: 0:00:00.027.                                                                                                                                                                                                                                                                                                                                                                                                       tools.py:301
           INFO     Preparing the tests ...                                                                                                                                                                                                                                                                                                                                                                                                                                tools.py:293
           INFO     Preparing the tests completed in: 0:00:00.001.                                                                                                                                                                                                                                                                                                                                                                                                         tools.py:301
           INFO     --- ANTA NRFU Run Information ---                                                                                                                                                                                                                                                                                                                                                                                                                     runner.py:276
                    Number of devices: 6 (6 established)                                                                                                                                                                                                                                                                                                                                                                                                                               
                    Total number of selected tests: 50                                                                                                                                                                                                                                                                                                                                                                                                                                 
                    Maximum number of open file descriptors for the current ANTA process: 16384                                                                                                                                                                                                                                                                                                                                                                                        
                    ---------------------------------                                                                                                                                                                                                                                                                                                                                                                                                                                  
           INFO     Preparing ANTA NRFU Run completed in: 0:00:00.033.                                                                                                                                                                                                                                                                                                                                                                                                     tools.py:301
           INFO     Running ANTA tests ...                                                                                                                                                                                                                                                                                                                                                                                                                                 tools.py:293
           INFO     Running ANTA tests completed in: 0:00:00.116.                                                                                                                                                                                                                                                                                                                                                                                                          tools.py:301
           INFO     Cache statistics for 'spine1': 0 hits / 8 command(s) (0.00%)                                                                                                                                                                                                                                                                                                                                                                                           runner.py:75
           INFO     Cache statistics for 'spine2': 0 hits / 8 command(s) (0.00%)                                                                                                                                                                                                                                                                                                                                                                                           runner.py:75
           INFO     Cache statistics for 'leaf1': 1 hits / 10 command(s) (10.00%)                                                                                                                                                                                                                                                                                                                                                                                          runner.py:75
           INFO     Cache statistics for 'leaf2': 1 hits / 10 command(s) (10.00%)                                                                                                                                                                                                                                                                                                                                                                                          runner.py:75
           INFO     Cache statistics for 'leaf3': 1 hits / 10 command(s) (10.00%)                                                                                                                                                                                                                                                                                                                                                                                          runner.py:75
           INFO     Cache statistics for 'leaf4': 1 hits / 10 command(s) (10.00%)                                                                                                                                                                                                                                                                                                                                                                                          runner.py:75
  • Running NRFU Tests...100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50/50 • 0:00:00 • 0:00:00

                                                    Summary per test                                                     
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Test Name                  ┃ # of success ┃ # of skipped ┃ # of failure ┃ # of errors ┃ List of failed or error nodes ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ VerifyBGPPeerCount         │ 0            │ 0            │ 0            │ 0           │                               │
├────────────────────────────┼──────────────┼──────────────┼──────────────┼─────────────┼───────────────────────────────┤
│ VerifyEOSVersion           │ 0            │ 0            │ 0            │ 0           │                               │
├────────────────────────────┼──────────────┼──────────────┼──────────────┼─────────────┼───────────────────────────────┤
│ VerifyInterfaceUtilization │ 0            │ 0            │ 0            │ 0           │                               │
├────────────────────────────┼──────────────┼──────────────┼──────────────┼─────────────┼───────────────────────────────┤
│ VerifyLoopbackCount        │ 0            │ 0            │ 0            │ 0           │                               │
├────────────────────────────┼──────────────┼──────────────┼──────────────┼─────────────┼───────────────────────────────┤
│ VerifyMlagInterfaces       │ 0            │ 0            │ 0            │ 0           │                               │
├────────────────────────────┼──────────────┼──────────────┼──────────────┼─────────────┼───────────────────────────────┤
│ VerifyMlagStatus           │ 0            │ 0            │ 0            │ 0           │                               │
├────────────────────────────┼──────────────┼──────────────┼──────────────┼─────────────┼───────────────────────────────┤
│ VerifyNTP                  │ 0            │ 0            │ 0            │ 0           │                               │
├────────────────────────────┼──────────────┼──────────────┼──────────────┼─────────────┼───────────────────────────────┤
│ VerifyReloadCause          │ 0            │ 0            │ 0            │ 0           │                               │
├────────────────────────────┼──────────────┼──────────────┼──────────────┼─────────────┼───────────────────────────────┤
│ VerifyZeroTouch            │ 0            │ 0            │ 0            │ 0           │                               │
└────────────────────────────┴──────────────┴──────────────┴──────────────┴─────────────┴───────────────────────────────┘
```

### PR DESCRIPTION
**Pull request by carl-baillargeon**
# Description

Fix ResultManager statistics. ResultManager stats are only updated when needed.

Fixes #948 

# Checklist:

- [x] Update documentation
- [x] Update unit tests

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -91,6 +91,7 @@ repos:
         name: Check typing with mypy
         args:
           - --config-file=pyproject.toml
+          - --explicit-package-bases
```
Comment (gmuloc) [inline .pre-commit-config.yaml:94]: can you explain why we need this?

File: anta/result_manager/__init__.py [hunk 2]
```diff
@@ -195,12 +264,15 @@ def add(self, result: TestResult) -> None:
...
+        # Register the manager to the result to allow status updates
+        result.register_manager(self)
+
         self._result_entries.append(result)
         self._update_status(result.result)
         self._update_stats(result)
 
         # Every time a new result is added, we need to clear the cached property
-        self.__dict__.pop("results_by_status", None)
+        self.clear_cache()
```
Comment (gmuloc) [inline anta/result_manager/__init__.py:275]: the way we use it it makes it a bit many calls

File: anta/result_manager/models.py [hunk 3]
```diff
@@ -58,6 +62,9 @@ class TestResult(BaseModel):
     messages: list[str] = []
     custom_field: str | None = None
 
+    # This is a circular import, so we need to use Optional
```
Comment (gmuloc) [inline anta/result_manager/models.py:65]: I don't feel like this is correct - where did you read about this?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (carl-baillargeon) [reply .pre-commit-config.yaml:94]: This is related to an issue we were having with mypy discovering imports --> https://anta.arista.com/main/contribution/#configure-mypypath

MYPYPATH used to do the trick but not anymore. Didn't investigate much. Here is the error I am getting:
```
Check typing with mypy...................................................Failed
- hook id: mypy
- exit code: 2

/home/cbaillar/git_projects/__init__.py: error: Source file found twice under different module names: "git_projects.anta.anta.logger" and "anta.logger"
/home/cbaillar/git_projects/__init__.py: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules for more info
/home/cbaillar/git_projects/__init__.py: note: Common resolutions include: a) adding `__init__.py` somewhere, b) using `--explicit-package-bases` or adjusting MYPYPATH
Found 1 error in 1 file (errors prevented further checking)
```

Can you test in your environment? I remember @titom73 and/or @mtache also had this issue in the past.

Adding this flag fixes it for now:P --> https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-paths-to-modules

File: anta/result_manager/models.py [hunk 4]
```diff
@@ -113,10 +120,25 @@ def _set_status(self, status: AntaTestStatus, message: str | None = None) -> Non
...
-        self.result = status
+        # Set the status via its ResultManager if available
+        if self._manager is not None:
+            self._manager.update_test_result(self, status, message)
+        else:
+            self.result = status
+            self.add_message(message)
+
+    def add_message(self, message: str | None = None) -> None:
+        """Add a message to the TestResult."""
```
[the remark this answers is not in the document]
Comment (carl-baillargeon) [reply anta/result_manager/models.py:134]: We are not setting results from the ResutManager so I think it's appropriate to have an exposed function for this.

File: anta/result_manager/models.py (hunk 3 above)
Comment (carl-baillargeon) [reply anta/result_manager/models.py]: I added a comment to explain why. Basically Ruff doesn't recognize private fields (with `_`) of Pydantic models and think types are needed at runtime. So it's either we ignore the Ruff error for the ResultManager import at the top or we use forward references.

File: tests/units/cli/nrfu/test_commands.py [hunk 5]
```diff
@@ -79,7 +79,7 @@ def test_anta_nrfu_text(click_runner: CliRunner) -> None:
 def test_anta_nrfu_text_multiple_failures(click_runner: CliRunner) -> None:
     """Test anta nrfu text with multiple failures, catalog is given via env."""
     result = click_runner.invoke(anta, ["nrfu", "text"], env={"ANTA_CATALOG": str(DATA_DIR / "test_catalog_double_failure.yml")})
-    assert result.exit_code == ExitCode.OK
+    assert result.exit_code == ExitCode.TESTS_FAILED
```
Comment (carl-baillargeon) [inline tests/units/cli/nrfu/test_commands.py:82]: @gmuloc That means it was broken before since it was not supposed to return ExitCode.OK right?

Comment (gmuloc) [reply tests/units/cli/nrfu/test_commands.py:82]: uh oh that looks bad but indeed it should be failed

File: .pre-commit-config.yaml (hunk 1 above)
Comment (gmuloc) [reply .pre-commit-config.yaml:94]: TBC - probably not needed

File: anta/result_manager/__init__.py (hunk 2 above)
Comment (gmuloc) [reply anta/result_manager/__init__.py:275]: lets use a flag

Comment (gmuloc) [reply anta/result_manager/__init__.py:275]: add a decorator on `result_by_status` -> if flag `stats_in_sync` is False then pop the porperty

Comment (carl-baillargeon) [reply anta/result_manager/__init__.py]: Fixed in 3c3e8ee3af4ff3ea57210e75835ee19d27f7105b

File: anta/result_manager/models.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (carl-baillargeon) [reply anta/result_manager/models.py]: Removed in 3c3e8ee3af4ff3ea57210e75835ee19d27f7105b

File: anta/result_manager/models.py (hunk 3 above)
Comment (carl-baillargeon) [reply anta/result_manager/models.py]: Removed in 3c3e8ee3af4ff3ea57210e75835ee19d27f7105b

File: anta/runner.py [hunk 6]
```diff
@@ -282,7 +283,7 @@ async def main(  # noqa: PLR0913
                 "Please consult the ANTA FAQ."
             )
 
-        coroutines = get_coroutines(selected_tests, manager)
+        coroutines = get_coroutines(selected_tests, manager if dry_run else None)
```
Comment (gmuloc) [inline anta/runner.py:286]: lets check if this was introduced after 1.1- remove manager from get_coroutine and do the same for loop in dry_run

Comment (gmuloc) [reply anta/runner.py:296]: ok this was added in 1.1 and was breaking .. so this fixes it technically

I think for v2.0.0 we need to think about take a decision on how the RM behaves (to quote @mtache maybe we don't want to have to give a RM to the runner)

Comment (carl-baillargeon) [reply anta/runner.py:296]: Correct. There is https://github.com/aristanetworks/anta/issues/837 tracking this.

**Review by gmuloc — APPROVED** [review]
LGTM - we can approve the codspeed impact

**Pull request merged**